File: packages/mobile/lib/providers/auth_provider.dart
import 'package:flutter/foundation.dart';
import 'package:flutter_secure_storage/flutter_secure_storage.dart';
import '../models/user.dart';
import '../services/api_service.dart';

class AuthProvider extends ChangeNotifier {
  User? _user;
  bool _isLoading = false;
  String? _error;

  User? get user => _user;
  bool get isAuthenticated => _user != null;
  bool get isLoading => _isLoading;
  String? get error => _error;

  // Demo users for all roles
  final Map<String, User> _demoUsers = {
    'admin@acm.com': User(
      id: '1',
      name: 'System Admin',
      email: 'admin@acm.com',
      role: UserRole.admin,
      permissions: ['all'],
    ),
    'analyst@acm.com': User(
      id: '2',
      name: 'SOC Analyst',
      email: 'analyst@acm.com',
      role: UserRole.analyst,
      permissions: [
        'view_alerts',
        'manage_alerts',
        'view_analytics',
        'generate_reports'
      ],
    ),
    'developer@acm.com': User(
      id: '3',
      name: 'API Developer',
      email: 'developer@acm.com',
      role: UserRole.developer,
      permissions: ['view_api_docs', 'manage_api_keys', 'view_analytics'],
    ),
    'viewer@acm.com': User(
      id: '4',
      name: 'Executive Viewer',
      email: 'viewer@acm.com',
      role: UserRole.viewer,
      permissions: ['view_dashboard', 'view_analytics'],
    ),
  };

  AuthProvider() {
    _loadSavedUser();
  }

  Future<void> _loadSavedUser() async {
    const storage = FlutterSecureStorage();
    final email = await storage.read(key: 'user_email');
    if (email != null && _demoUsers.containsKey(email)) {
      _user = _demoUsers[email];
      notifyListeners();
    }
  }

  Future<bool> login(String email, String password) async {
    _isLoading = true;
    _error = null;
    notifyListeners();

    try {
      // Demo authentication
      await Future.delayed(const Duration(milliseconds: 500));

      final normalizedEmail = email.toLowerCase();
      if (_demoUsers.containsKey(normalizedEmail) && password == 'demo123') {
        _user = _demoUsers[normalizedEmail];

        // Save to secure storage
        const storage = FlutterSecureStorage();
        await storage.write(key: 'user_email', value: normalizedEmail);

        // Generate demo token
        final token = 'demo_token_${DateTime.now().millisecondsSinceEpoch}';
        await ApiService.setToken(token);

        _isLoading = false;
        notifyListeners();
        return true;
      } else {
        _error = 'Invalid email or password';
        _isLoading = false;
        notifyListeners();
        return false;
      }
    } catch (e) {
      _error = 'An error occurred. Please try again.';
      _isLoading = false;
      notifyListeners();
      return false;
    }
  }

  Future<void> logout() async {
    _user = null;
    _error = null;

    const storage = FlutterSecureStorage();
    await storage.delete(key: 'user_email');
    await ApiService.clearToken();

    notifyListeners();
  }

  bool hasPermission(String permission) {
    return _user?.hasPermission(permission) ?? false;
  }

  bool hasRole(List<UserRole> roles) {
    return _user?.hasRole(roles) ?? false;
  }

  void clearError() {
    _error = null;
    notifyListeners();
  }
}

File: voxguard/frontend/src/providers/auth-provider.ts
import type { AuthProvider } from '@refinedev/core';

// User interface matching Hasura schema
export interface User {
  id: string;
  email: string;
  name: string;
  role: 'admin' | 'analyst' | 'developer' | 'viewer';
  avatar?: string;
  permissions: string[];
  created_at: string;
  last_login?: string;
}

// Demo users for development (will be replaced by Hasura auth)
const DEMO_USERS: Record<string, { password: string; user: User }> = {
  'admin@acm.com': {
    password: 'demo123',
    user: {
      id: '1',
      email: 'admin@acm.com',
      name: 'Admin User',
      role: 'admin',
      permissions: ['all'],
      created_at: '2024-01-01T00:00:00Z',
    },
  },
  'analyst@acm.com': {
    password: 'demo123',
    user: {
      id: '2',
      email: 'analyst@acm.com',
      name: 'Security Analyst',
      role: 'analyst',
      permissions: ['alerts:read', 'alerts:update', 'analytics:read'],
      created_at: '2024-01-01T00:00:00Z',
    },
  },
  'developer@acm.com': {
    password: 'demo123',
    user: {
      id: '3',
      email: 'developer@acm.com',
      name: 'Developer',
      role: 'developer',
      permissions: ['alerts:read', 'analytics:read', 'settings:read'],
      created_at: '2024-01-01T00:00:00Z',
    },
  },
  'viewer@acm.com': {
    password: 'demo123',
    user: {
      id: '4',
      email: 'viewer@acm.com',
      name: 'Viewer',
      role: 'viewer',
      permissions: ['alerts:read', 'analytics:read'],
      created_at: '2024-01-01T00:00:00Z',
    },
  },
};

// Storage key for auth data
const AUTH_STORAGE_KEY = 'voxguard-auth';

// Generate a mock JWT token (in production, this comes from Hasura/Auth service)
const generateToken = (user: User): string => {
  const header = btoa(JSON.stringify({ alg: 'HS256', typ: 'JWT' }));
  const payload = btoa(
    JSON.stringify({
      sub: user.id,
      email: user.email,
      role: user.role,
      'https://hasura.io/jwt/claims': {
        'x-hasura-allowed-roles': [user.role],
        'x-hasura-default-role': user.role,
        'x-hasura-user-id': user.id,
      },
      iat: Math.floor(Date.now() / 1000),
      exp: Math.floor(Date.now() / 1000) + 86400, // 24 hours
    })
  );
  const signature = 'mock-signature';
  return `${header}.${payload}.${signature}`;
};

// Get stored auth data
const getStoredAuth = (): { user: User; token: string } | null => {
  try {
    const stored = localStorage.getItem(AUTH_STORAGE_KEY);
    if (stored) {
      return JSON.parse(stored).state;
    }
  } catch (e) {
    console.error('Error reading auth storage:', e);
  }
  return null;
};

// Store auth data
const storeAuth = (user: User, token: string): void => {
  localStorage.setItem(
    AUTH_STORAGE_KEY,
    JSON.stringify({
      state: { user, token, isAuthenticated: true },
    })
  );
};

// Clear auth data
const clearAuth = (): void => {
  localStorage.removeItem(AUTH_STORAGE_KEY);
};

// Refine Auth Provider
export const authProvider: AuthProvider = {
  login: async ({ email, password }) => {
    // Check demo users
    const demoUser = DEMO_USERS[email];
    if (demoUser && demoUser.password === password) {
      const token = generateToken(demoUser.user);
      storeAuth(demoUser.user, token);

      return {
        success: true,
        redirectTo: '/dashboard',
      };
    }

    // TODO: Implement real Hasura authentication
    // const response = await fetch('/api/auth/login', {
    //   method: 'POST',
    //   headers: { 'Content-Type': 'application/json' },
    //   body: JSON.stringify({ email, password }),
    // });
    // const data = await response.json();

    return {
      success: false,
      error: {
        name: 'LoginError',
        message: 'Invalid email or password',
      },
    };
  },

  logout: async () => {
    clearAuth();
    return {
      success: true,
      redirectTo: '/login',
    };
  },

  check: async () => {
    const auth = getStoredAuth();
    if (auth?.token) {
      return {
        authenticated: true,
      };
    }

    return {
      authenticated: false,
      redirectTo: '/login',
      error: {
        name: 'Unauthorized',
        message: 'Please login to continue',
      },
    };
  },

  getPermissions: async () => {
    const auth = getStoredAuth();
    if (auth?.user) {
      return auth.user.permissions;
    }
    return [];
  },

  getIdentity: async () => {
    const auth = getStoredAuth();
    if (auth?.user) {
      return {
        id: auth.user.id,
        name: auth.user.name,
        email: auth.user.email,
        avatar: auth.user.avatar || `https://ui-avatars.com/api/?name=${encodeURIComponent(auth.user.name)}&background=1B4F72&color=fff`,
        role: auth.user.role,
      };
    }
    return null;
  },

  onError: async (error) => {
    if (error?.statusCode === 401 || error?.statusCode === 403) {
      return {
        logout: true,
        redirectTo: '/login',
        error,
      };
    }

    return { error };
  },
};

export default authProvider;
